nat/traversal/pinger.go

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package traversal
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
+	"golang.org/x/net/ipv4"
+)
+
+const prefix = "[NATPinger] "
+const pingInterval = 200
+const pingTimeout = 10000
+
+// ConfigParser is able to parse a config from given raw json
+type ConfigParser interface {
+	Parse(config json.RawMessage) (ip string, port int, err error)
+}
+
+// Pinger represents NAT pinger structure
+type Pinger struct {
+	localPort      int
+	pingTarget     chan json.RawMessage
+	pingReceived   chan struct{}
+	pingCancelled  chan struct{}
+	natEventWaiter NatEventWaiter
+	configParser   ConfigParser
+	stop           chan struct{}
+	once           sync.Once
+}
+
+// NatEventWaiter is responsible for waiting for nat events
+type NatEventWaiter interface {
+	WaitForEvent() Event
+}
+
+// NewPingerFactory returns Pinger instance
+func NewPingerFactory(waiter NatEventWaiter, parser ConfigParser) *Pinger {
+	target := make(chan json.RawMessage)
+	received := make(chan struct{})
+	cancel := make(chan struct{})
+	stop := make(chan struct{})
+	return &Pinger{
+		pingTarget:     target,
+		pingReceived:   received,
+		pingCancelled:  cancel,
+		natEventWaiter: waiter,
+		configParser:   parser,
+		stop:           stop,
+	}
+}
+
+// Start starts NAT pinger and waits for pingTarget to ping
+func (p *Pinger) Start() {
+	log.Info(prefix, "Starting a NAT pinger")
+	for {
+		select {
+		case <-p.stop:
+			return
+		case dst := <-p.pingTarget:
+			IP, port, err := p.configParser.Parse(dst)
+			if err != nil {
+				log.Warn(prefix, errors.Wrap(err, fmt.Sprintf("unable to parse ping message: %v", string(dst))))
```

Comment:
We are not returning the error here, just printing:
```suggestion
				log.Warn(prefix, fmt.Sprintf("unable to parse ping message: %v: %v", string(dst), err))
```